### ISSUE TITLE
slower and less accurate performance on a simple example relative to sklearn

### ISSUE THREAD

**Issue by chananshgong** [jmschrei/pomegranate#240]
This is just a simple test which shows performance is less accurate and slower with pomegrante GMM relative to SKLEARN (before taking into account HMM):

```
model = GaussianMixture (sklearn)
=========================
score = 1.0
time elapsed = 0.0194020271301

model = GeneralMixtureModel (Pomegrante)
=========================
score = 0.813037310621
time elapsed = 0.101740837097
```

```
"""test the supervised gmm classes in sklearn vs pomegranate"""
import numpy as np
from hmmlearn import hmm
import warnings
import unittest
from sklearn.metrics.cluster.supervised import homogeneity_score
import time
from sklearn.mixture import GaussianMixture
from pomegranate.gmm import GeneralMixtureModel
from pomegranate.distributions import MultivariateGaussianDistribution
class Test(unittest.TestCase):
    def setUp(self):
        np.random.seed(42)
        n_states = 4
        n_samples = int(1e4)
        # Prepare parameters for a 4-components HMM, taken from: http://hmmlearn.readthedocs.io/en/latest/auto_examples/plot_hmm_sampling.html#sphx-glr-auto-examples-plot-hmm-sampling-py. It's an overkill for this example. It's just for proving the point.
        # Initial population probability
        startprob = np.array([0.6, 0.3, 0.1, 0.0])
        # The transition matrix, note that there are no transitions possible
        # between component 1 and 3
        transmat = np.array([[0.7, 0.2, 0.0, 0.1],
                             [0.3, 0.5, 0.2, 0.0],
                             [0.0, 0.3, 0.5, 0.2],
                             [0.2, 0.0, 0.2, 0.6]])
        # The means of each component
        means = np.array([[0.0, 0.0],
                          [0.0, 11.0],
                          [9.0, 10.0],
                          [11.0, -1.0]])
        # The covariance of each component
        covars = .5 * np.tile(np.identity(2), (4, 1, 1))

        # Build an HMM instance and set parameters
        model = hmm.GaussianHMM(n_components=4, covariance_type="full")
        # Instead of fitting it from the data, we directly set the estimated
        # parameters, the means and covariance of the components
        model.startprob_ = startprob
        model.transmat_ = transmat
        model.means_ = means
        model.covars_ = covars
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore", category=DeprecationWarning)
            x, y = model.sample(n_samples)
        self.x = x
        self.y = y
        self.n_states = n_states

    def test_gmm(self):
        models = [GaussianMixture(n_components=self.n_states, covariance_type='full'),
                  GeneralMixtureModel(MultivariateGaussianDistribution, n_components=self.n_states)]
        x = self.x
        y = self.y
        for model in models:
            np.random.seed(42)
            print('model = %s\n=========================' % model.__class__.__name__)
            t0 = time.time()
            model.fit(x)
            yhat = model.predict(x)
            t1 = time.time()
            t_elapsed = t1 - t0
            score = homogeneity_score(y, yhat)
            print('score = %s' % score)
            print('time elapsed = %s' % t_elapsed)
            self.assertGreater(score, 0.9)
            print('')


if __name__ == '__main__':
    unittest.main()
```

**Comment by jmschrei** [comment]
Howdy @chananshgon. Sorry for the late reply. 

The issue is that pomegranate uses kmeans and not kmeans++ as the initialization step for GMM and so sometimes that produces bad results. That seed, 42, just happens to produce bad results. Basically, since your clusters are linearly separable it should take only the initialization step to properly fit the GMM. However, if you have a bad initialization, such as splitting one of the clusters in two, then the EM algorithm is run for the default number of iterations trying to improve the bad result. I modified your code a bit but it should become apparent if you run it what is happening. When both work properly (which happens the majority of time), sklearn takes ~0.24 seconds and pomegranate takes ~0.15 seconds on my computer.

```
"""test the supervised gmm classes in sklearn vs pomegranate"""
import numpy as np
from hmmlearn import hmm
import warnings
import unittest
from sklearn.metrics.cluster.supervised import homogeneity_score
import time
from sklearn.mixture import GaussianMixture
from pomegranate.gmm import GeneralMixtureModel
from pomegranate.distributions import MultivariateGaussianDistribution

import matplotlib.pyplot as plt


#np.random.seed(42)
n_states = 4
n_samples = int(1e4)
# Prepare parameters for a 4-components HMM, taken from: http://hmmlearn.readthedocs.io/en/latest/auto_examples/plot_hmm_sampling.html#sphx-glr-auto-examples-plot-hmm-sampling-py. It's an overkill for this example. It's just for proving the point.
# Initial population probability
startprob = np.array([0.6, 0.3, 0.1, 0.0])
# The transition matrix, note that there are no transitions possible
# between component 1 and 3
transmat = np.array([[0.7, 0.2, 0.0, 0.1],
                     [0.3, 0.5, 0.2, 0.0],
                     [0.0, 0.3, 0.5, 0.2],
                     [0.2, 0.0, 0.2, 0.6]])
# The means of each component
means = np.array([[0.0, 0.0],
                  [0.0, 11.0],
                  [9.0, 10.0],
                  [11.0, -1.0]])
# The covariance of each component
covars = .5 * np.tile(np.identity(2), (4, 1, 1))

# Build an HMM instance and set parameters
model = hmm.GaussianHMM(n_components=4, covariance_type="full")
# Instead of fitting it from the data, we directly set the estimated
# parameters, the means and covariance of the components
model.startprob_ = startprob
model.transmat_ = transmat
model.means_ = means
model.covars_ = covars
with warnings.catch_warnings():
    warnings.filterwarnings("ignore", category=DeprecationWarning)
    X, y = model.sample(n_samples)


model = GaussianMixture(n_components=n_states, covariance_type='full', tol=0.1, max_iter=10, verbose=2, verbose_interval=1)

print "GaussianMixture"
tic = time.time()
model.fit(X)
print "Fitting Time: ", time.time() - tic

tic = time.time()
yhat = model.predict(X)
print "Prediction Time: ", time.time() - tic
print "Score: ", homogeneity_score(y, yhat)
print "\n\n"

x1 = X[yhat == 0]
x2 = X[yhat == 1]
x3 = X[yhat == 2]
x4 = X[yhat == 3]

plt.scatter(x1[:,0], x1[:,1], color='c')
plt.scatter(x2[:,0], x2[:,1], color='r')
plt.scatter(x3[:,0], x3[:,1], color='m')
plt.scatter(x4[:,0], x4[:,1], color='g')
plt.show()


model = GeneralMixtureModel(MultivariateGaussianDistribution, n_components=n_states)

print "GeneralMixtureModel"
tic = time.time()
model.fit(X, max_iterations=100, stop_threshold=0.1, verbose=True)
print "Fitting Time: ", time.time() - tic

tic = time.time()
yhat = model.predict(X)
print "Prediction Time: ", time.time() - tic
print "Score: ", homogeneity_score(y, yhat)

x1 = X[yhat == 0]
x2 = X[yhat == 1]
x3 = X[yhat == 2]
x4 = X[yhat == 3]

plt.scatter(x1[:,0], x1[:,1], color='c')
plt.scatter(x2[:,0], x2[:,1], color='r')
plt.scatter(x3[:,0], x3[:,1], color='m')
plt.scatter(x4[:,0], x4[:,1], color='g')
plt.show()
```

The solution to this is probably to implement kmeans++ in pomegranate. However, this is not a bug in pomegranate, as much as it is an unfortunate reality of the vanilla kmeans initialization.

**Comment by chananshgong** [comment]
Thanks for the clarification.

**Comment by jkleckner** [comment]
Note that [wikipedia](https://en.wikipedia.org/wiki/K-means%2B%2B#Applications) mentions the [k-meansII](http://vldb.org/pvldb/vol5/p622_bahmanbahmani_vldb2012.pdf) algorithm as having the k-means++ guarantees but being more scalable.

Is there some way to work around the initialization problem, other than just trying different random seeds?

**Comment by jmschrei** [comment]
Howdy @jkleckner , thanks for the links! sklearn recently attempted to implement kmeans||, but found that it didn't seem to actually perform better: https://github.com/scikit-learn/scikit-learn/pull/8585. It is possible that there was an issue in the implementation though.

The most traditional way is to run the algorithm multiple times and take the best initialization. I think that I can take a stab at either kmeans|| or kmeans++ as alternate initializations, but given that my queue has gotten pretty long recently I'm not sure when I will get around to it. Note that if you'd like to use someone elses implementations, you can always feed in the initial distributions that the other algorithm found and fit from there.

**Comment by jmschrei** [comment]
So I went ahead and added a fast implementation of kmeans++, made it the default, and made kmeans default to re-initializing 10 times and providing the best result (`n_init` as the input parameter). This mirrors scikit-learn's defaults. Doing this seems to yield results just as good as scikit-learn, as one would expect.

I also attempted to add in kmeans||. It seems as though kmeans++ performs well when you have a smaller number of clusters (maybe 10-20), but kmeans|| performs bettwe when you have a larger number of clusters and a larger amount of data (50+ clusters, 50,000+ samples). I haven't done a thorough analysis of faster vs better and counting iterations vs initialization time, but I thought I'd let y'alls know that kmeans++ and kmeans|| are now both available for kmeans and GMM through the `init` keyword and will produce better results.

Let me know if there are any issues with either of them!
